### PR TITLE
Separate GyroBalancer drive control loop from balance loop

### PR DESCRIPTION
**Pull request by gregcowell**
Please see issue #429. This facilitates the writing of simple scripts for controlling the movement of a GyroBalancer robot (eg.BALANC3R).

I recommend reviewing the new example scripts [here](https://github.com/ev3dev/ev3dev-lang-python-demo/pull/26) first. 

This pull request also adds the ability to enable debug mode. In debug mode, the GyroBalancer robot logs gyro reading and loop time statistics.

### CODE REVIEW

File: ev3dev/control/GyroBalancer.py [hunk 1]
```diff
@@ -29,44 +29,48 @@
...
+from ev3dev.sound import Sound
+from collections import OrderedDict
 
 log = logging.getLogger(__name__)
 
-########################################################################
-# File I/O functions
-########################################################################
+# Constants
+RAD_PER_DEG = 3.14159265/180
```
Comment (dwalton76) [inline ev3dev/control/GyroBalancer.py]: I’m not sure how many decimal places it goes but you could do “from math import pi” and use “pi” here

File: ev3dev/control/GyroBalancer.py [hunk 2]
```diff
@@ -29,44 +29,48 @@
...
+# Express the above as the angle in rad per motor unit
+RAD_PER_RAW_MOTOR_UNIT = DEG_PER_RAW_MOTOR_UNIT * RAD_PER_DEG
 
-def SetDuty(motorDutyFileHandle, duty, frictionOffset, voltageCompensation):
-    """Function to set the duty cycle of the motors."""
-    # Compensate for nominal voltage and round the input
-    dutyInt = int(round(duty*voltageCompensation))
+# On the EV3, "1% speed" corresponds to 1.7 RPM (if speed
+# control were enabled).
+RPM_PER_PERCENT_SPEED = 1.7
```
Comment (dwalton76) [inline ev3dev/control/GyroBalancer.py:65]: Doesn’t this assume LargeMotor is in use? Maybe make this dynamic based on the motor class that is being used.

File: ev3dev/control/GyroBalancer.py [hunk 3]
```diff
@@ -100,383 +105,379 @@ def __init__(self,
...
-                log.info("-----------------------------------")
-                log.info("STOP")
-                log.info("-----------------------------------")
-
-        # Exit cleanly so that all motors are stopped
-        except (KeyboardInterrupt, Exception) as e:
-            log.exception(e)
-            shutdown()
+        self.motor_left = LargeMotor(left_motor_port)
+        self.motor_right = LargeMotor(right_motor_port)
```
Comment (dwalton76) [inline ev3dev/control/GyroBalancer.py:133]: Allow the use to pass in the motor class to use but have it default to LargeMotor? I can’t think of any balancers that use medium motors but hey maybe we could build one :)

**Review by dwalton76 — CHANGES_REQUESTED** [review]
Would be handy to catch SIGTERM and SIGINT and have those set stop_balance. I have an example of this (it uses ‘signal’) in some of the demo robots....MINDCUBER and probably GRIPPER will do it.

**Comment by dwalton76** [comment]
PS
Supporting motors other than LargeMotor isn’t super important to me, feel free to ignore those comments if you want.

**Comment by gregcowell** [comment]
>Would be handy to catch SIGTERM and SIGINT and have those set stop_balance.

Good idea. I've added the relevant signal handlers. On a related note, I suspect that Brickman sends a SIGKILL when the back button is pressed and we can't have a handler for that. 

>you could do “from math import pi” and use “pi” here

Yep, done.

>Supporting motors other than LargeMotor isn’t super important to me, feel free to ignore those comments if you want.

Thanks @dwalton76. I think I'll leave it with just LargeMotor support at least for now.

**Comment by WasabiFan** [comment]
@dwalton76 Can you take a quick look at this and confirm it's good to go?

**Review by dwalton76 — APPROVED** [review]

File: ev3dev/control/GyroBalancer.py (hunk 2 above)
Comment (laurensvalk) [reply ev3dev/control/GyroBalancer.py:65]: Yes. the factor is 2.67 for medium motor.

**Pull request merged**